src/account.ts

```diff
@@ -1,23 +1,124 @@
- */
-export const zeroAddress = function(): string {
-  const addressLength = 20
-  const addr = zeros(addressLength)
-  return bufferToHex(addr)
+export interface AccountData {
+  nonce?: BNLike
+  balance?: BNLike
+  stateRoot?: BufferLike
+  codeHash?: BufferLike
+}
+
+type BNLike = BN | string | number
+
+type BufferLike = Buffer | TransformableToBuffer | PrefixedHexString | number
+
+type PrefixedHexString = string
+
+interface TransformableToBuffer {
+  toBuffer(): Buffer
+}
+
+/* Helper for Account.serialize() */
+function bnToRlp(value: BN): Buffer {
+  // using `bn.toArrayLike(Buffer)` instead of `bn.toBuffer()`
+  // for compatibility with browserify and similar tools
+  return unpadBuffer(value.toArrayLike(Buffer))
+}
+
+export class Account {
+  nonce: BN
+  balance: BN
+  stateRoot: Buffer
+  codeHash: Buffer
+
+  static fromAccountData(accountData: AccountData) {
+    const { nonce, balance, stateRoot, codeHash } = accountData
+
+    return new Account(
+      nonce ? new BN(toBuffer(nonce)) : undefined,
+      balance ? new BN(toBuffer(balance)) : undefined,
+      stateRoot ? toBuffer(stateRoot) : undefined,
+      codeHash ? toBuffer(codeHash) : undefined,
+    )
+  }
+
+  public static fromRlpSerializedAccount(serialized: Buffer) {
+    const values = rlp.decode(serialized)
+
+    if (!Array.isArray(values)) {
+      throw new Error('Invalid serialized account input. Must be array')
+    }
+
+    return this.fromValuesArray(values)
+  }
+
+  public static fromValuesArray(values: Buffer[]) {
+    const [nonce, balance, stateRoot, codeHash] = values
+
+    return new Account(
+      nonce ? new BN(nonce) : undefined,
+      balance ? new BN(balance) : undefined,
+      stateRoot,
+      codeHash,
+    )
+  }
+
+  /**
+   * This constructor takes the values, validates and assigns them.
+   * Use the static factory methods to assist in creating an Account from varying data types.
+   */
+  constructor(
+    nonce = new BN(0),
+    balance = new BN(0),
+    stateRoot = KECCAK256_RLP,
+    codeHash = KECCAK256_NULL,
+  ) {
+    if (stateRoot.length !== 32) {
+      throw new Error('stateRoot must have a length of 32')
+    }
+    if (codeHash.length !== 32) {
+      throw new Error('codeHash must have a length of 32')
+    }
+
+    this.nonce = nonce
+    this.balance = balance
+    this.stateRoot = stateRoot
+    this.codeHash = codeHash
+  }
+
+  /**
+   * Returns the RLP serialization of the account as a `Buffer`.
+   */
+  serialize(): Buffer {
+    return rlp.encode([bnToRlp(this.nonce), bnToRlp(this.balance), this.stateRoot, this.codeHash])
+  }
+
+  /**
+   * Returns a `Boolean` deteremining if the account is a contract.
+   */
+  isContract(): boolean {
+    return !this.codeHash.equals(KECCAK256_NULL)
+  }
+
+  /**
+   * Returns a `Boolean` determining if the account is empty.
+   */
+  isEmpty(): boolean {
+    return this.balance.isZero() && this.nonce.isZero() && this.codeHash.equals(KECCAK256_NULL)
```

Comment:
I saw this previously in our monorepo too but forgot to raise this. I think we should also explicitly add that `this.stateRoot.equals(KECCAK256_RLP)`. 

The reason for this is that in the current HF it is not possible to instantiate a contract which would return empty (nonce is set to 1 if the tx succeeds). However, in at least Frontier nonce is not set to 1 and instead remains 0 upon creation. It is thus possible to create a contract where in the constructor you store some values in storage (thus `stateRoot` is not `KECCAK256_RLP`), but you do not put any code in. So in that case you have created a non-empty account which this function would currently say is empty. I think this might lead to some problems if we ever want to full sync mainnet.

Reply:
good idea, added in https://github.com/ethereumjs/ethereumjs-util/pull/275/commits/cfe8a9a13d94a4d2c9ac05d2e590261f7b9a3f5b